lib/path.dart

```diff
@@ -0,0 +1,22 @@
+import 'dart:html' as html;
+
+const LOCATION_EVENT_NAME = 'locationChanged';
+
+String get currentName => html.window.location.pathname;
+
+void goto(String pathname, {bool replace = false}) {
+  if (html.window.location.pathname == pathname) return;
+
+  if (replace) {
+    html.window.history.replaceState({}, '', pathname);
+  } else {
+    html.window.history.pushState({}, '', pathname);
+  }
+
+  html.window.dispatchEvent(html.Event(LOCATION_EVENT_NAME));
```

Comment:
I'm not sure it's a good idea to add indirection via custom events (unless it's really necessary) - I find it makes it much harder to trace the code when something goes wrong. It's less problematic here as it's a pretty straightforward app, but can we think of another way of achieving this? Perhaps the `App` constructor can just set a variable in this file that specifies the function to be called at the end of the `goto` function?

I'm also not adverse to not having a separate `path.dart` file, and just having all of the routing code in the `app.dart` instead, with the navbar controller importing `goto` from `app.dart` if it makes it more straightforward. Adding layers of abstraction is far too overrated - better a more simple and direct solution, even if more tightly coupled.